functional_tests/steps/auth.py

```diff
@@ -0,0 +1,204 @@
+    helper = AuthenticationTestHelper(context)
+    helper.setup_test_keypair()
+    helper.generate_test_tokens(groups=["role-admin"])
+    cookies = helper.create_auth_cookies()
+    context.page.context.add_cookies(cookies)
+    helper.setup_session_renewal_timing()
+
+
+@when("the user navigates to the admin page")
+def navigate_to_admin(context: Context) -> None:
+    """Navigate to the Wagtail admin page and simulate user activity."""
+    # Navigate to admin
+    context.page.goto(f"{context.base_url}/admin/")
+
+
+@when("the user becomes active again")
+@when("the user is active in the admin interface")
+@when("the JWT token is refreshed in one tab")
+def simulate_user_activity(context: Context) -> None:
+    """Simulate mouse movements to trigger activity detection."""
+    movements = [(500, 300), (600, 350), (500, 300)]
+
+    for x, y in movements:
+        context.page.mouse.move(x, y)
+        context.page.wait_for_timeout(4000)
+
+    context.page.wait_for_timeout(1000)  # Allow time for any session renewal logic to run
+
+
+@then("a session renewal request should be sent")
+def verify_renewal_requested(context: Context) -> None:
+    auth_config = get_auth_config()
+    refresh_path = auth_config["authTokenRefreshUrl"]
+
+    require_request(
+        context._requests,  # pylint: disable=protected-access
+        lambda r: "/extend-session/" in r.url and r.method in ("POST", "PUT"),
+        "extend_session view call",
+    )
+    require_request(
+        context._requests,  # pylint: disable=protected-access
+        lambda r: refresh_path in r.url and r.method in ("POST", "PUT"),
+        f"renewal request to {refresh_path}",
+    )
+
+
+@given("the user has no valid JWT tokens")
+def step_clear_tokens(context: Context) -> None:
+    """Clear any existing auth cookies (no valid tokens)."""
+    context.page.context.clear_cookies()
+
+
+@then("the logout request should complete successfully")
+@then("the user should be redirected to the sign-in page")
+def step_redirected_to_signin(context: Context) -> None:
+    """Verify we were redirected to the admin login page."""
+    expected_path = "/admin/login/"
+    current_url = context.page.url
+    if expected_path not in current_url:
+        raise AssertionError(f"Not redirected to login, current URL: {current_url}")
+
+
+@then("the user should not be redirected to the sign-in page")
+def step_not_redirected_to_signin(context: Context) -> None:
+    """Verify we stayed on admin and did not hit the login URL."""
+    forbidden = "/admin/login/"
+    current = context.page.url
+    if forbidden in current:
+        raise AssertionError(f"Unexpected redirect to login page: {current}")
+
+
+@when("the user remains inactive for a period longer than the token's expiration time")
+def step_wait_for_expiry(context: Context) -> None:
+    """Sleep past the token TTL so it expires."""
+    context.page.wait_for_timeout(20000)
```

Comment:
```suggestion
    context.page.wait_for_timeout(1000)
```
This slows down the tests more than necessary, the 20 seconds is fine where we're doing comparison of the value rather than actually waiting the specified time. We should be ok to just change the `"exp"` field in the payload generated in `build_jwt`.

Reply:
question: could the expiry be set lower globally or only for this test?
The functional tests already run long, and I would like us to reduce long wait as much as possible

Reply:
I already had things in place from my `helpers.py` from my unit/ integration test helpers. So now I am utilising those with some tweaks within `auth_utils.py`

Reduce the wait from `20000` 20 seconds to `5000` 5 seconds, while keeping the tests stable/ no flakiness and having it reliable. Originally, I increased the time to reduce flakiness and reliability, capturing the logs and some tests taking longer to run, such as the multiple tabs one.

[8879f65](https://github.com/ONSdigital/dis-wagtail/pull/338/commits/8879f6569797202be84435b5363f6b69b7269edd)

@kacperpONS